### ISSUE TITLE
Upgrade vagrant box and/or docs?

### ISSUE THREAD

**Issue by eiriksm** [npm/newww#476]
Hello.

While preparing for submitting a patch today, I encountered a series of hurdles which I had to overcome to get the local development up and running. This issue will serve as a summary of that, and I'd be happy to create separate issues (and PRs) for things that needs to be fixed. This is for discussion.

Let me just say that my motivation for writing this is that I think it is awesome that we can all contribute to the npm website, and it is awesome that a vagrant box exists to get up and running from everywhere. And of course npm is on top of the Niceness Per Minute(tm) scale. My motivation is to make the first boot experience great too.

First:
when following the exact steps in the readme, the web site does not start. In fact, it does not install. This is probably the same issue as @nfriedly is having over at #441.

A little snip from the log is this:
```
> hiredis@0.1.17 install /vagrant/node_modules/hiredis
> node-gyp rebuild

make: Entering directory `/vagrant/node_modules/hiredis/build'
  CC(target) Release/obj.target/hiredis/deps/hiredis/hiredis.o
  CC(target) Release/obj.target/hiredis/deps/hiredis/net.o
  CC(target) Release/obj.target/hiredis/deps/hiredis/sds.o
  CC(target) Release/obj.target/hiredis/deps/hiredis/async.o
  AR(target) Release/obj.target/deps/hiredis.a
  COPY Release/hiredis.a
  CXX(target) Release/obj.target/hiredis/src/hiredis.o
make: g++: Command not found
```

OK, so no g++. This is solved by simply running `sudo apt-get install g++`. 

Also, node version is 0.10.26. Which is not the most recent one :p

There were other small problems as well, but this brings me to my points:
- I have been working alot with vagrant, and would love to add these fixes to the vagrant box. But where would I post PRs to the box? And how often does it get built and posted?
- I could just add some extra lines to the Vagrantfile to make sure node is updated, along with installing g++. But then we are updating an outdated version of node instead of just installing the newest one. Which is preferred?
- I could add some lines in the docs about "install g++" and so on, but that feels almost like excuses for not doing it right in the first place. If you see what I mean.

Ok. That's a start. Let's make the "get up and running" with the npm website feel like "No Problems Met"(tm)

Have a great day!

**Comment by zeke** [comment]
Thanks for taking the time to write up this issues. Here's the TLDR:

- A compiler is required
- Node needs to be updated
- Readme should reflect this stuff

I'll defer to @seldo and @rockbot for more help on this. They're the VM experts. :)

**Comment by seldo** [comment]
The point of the VM is that should be a one-click setup process; the best course of action is to update the VM (or create a new VM that is compatible) and publish a new .box file that's got the current version of node and the compiler already installed. Until now that's been my job but if you'd like to take a crack at it that would be great, too :-)

**Comment by zeke** [comment]
Maybe the new box should use iojs, which is supposed to ship next week. :)

**Comment by eiriksm** [comment]
This might be just me, and maybe that is the reason I opened the issue and therefore it is not an issue, but I feel like the box should not be a .box

How I work with vagrant is having all install and config in the repo. Or a separete repo. This way people can change it around in PRs, and upgrading node would not even be an issue (since a bootstrap script would install the latest node).

I appreciate that the npm site is a complex site - and that making stuff install in a bootstrap script makes the initial state more in flux, but still... Something to consider?

As to the the setup process being one-click, imo it would not be less one-click. In fact, I would almost argue it would be even faster than the current one, since we could base it on a standard ubuntu (for example) vagrant box, which a lot of people are a lot more likely to already have downloaded. As opposed to a npm specific .box, which almost always will be a new download to new contributors.

Sorry if that was a bit forward and opinionated. But going with the initial style of the first post, these are just my impressions as I decided to try to make my first PR a while back. And therefore i just assume some others might feel the same way :)

Ok, so TDLR from this long comment:
- I (personally) prefer boxes built with scripts (that you can send PRs to)
- This would solve all issues listed in the above tldr by @zeke (actually it would even make it trivial switching to iojs ;))
- I am more than happy to contribute to make it simpler.
- I am also happy to help going forward with the current .box "format", so let me know if I can help there.

Btw, have a great friday everyone! :)

**Comment by seldo** [comment]
I appreciate your position here, and usually that's how I would structure my Vagrant boxes, but in this case getting Couch installed is kind of an involved process that isn't very amenable to automation. Once I had to have a binary of some kind, it made sense to just throw everything into the binary.

Of course, it could just be that my vagrant-foo is weak. If you make a Vagrantfile that successfully bootstraps the www image from a base Vagrant image I would love that.

**Comment by eiriksm** [comment]
Sounds like a fun challenge for a saturday morning :)

I'll see if I can make that happen

**Comment by eiriksm** [comment]
Hey, so I have this working now.

But seems to me that in the dev box included in the repo couchdb is 1.5, not 1.6. You want to update it?
Also, elasticsearch is on 1.1, should I keep it at that version?
Ubuntu is 12.04, seems like it should be 14.04?

IMO these should reflect the live servers in the best possible way. If it is not rude to ask: Are you running these versions of couch, elasticsearch and ubuntu there?

So, pending these questions...PR coming up as promised :)

Have a nice weekend!

**Comment by eiriksm** [comment]
Opened a PR instead (#524), so we can continue the discussion over there. So closing this. 

Feel free to reopen if your workflow suggests otherwise. :)
